### ISSUE TITLE
Add dirsync support

### ISSUE THREAD

**Issue by behlendorf** [zfsonlinux/zfs#4366]
Add support for the Linux `dirsync` mount option as described in the `mount(8)` man page.

```
       dirsync
              All  directory updates within the filesystem should be done syn‐
              chronously.  This affects the  following  system  calls:  creat,
              link, unlink, symlink, mkdir, rmdir, mknod and rename.
```

It would be nice to support this standard Linux mount option.  The most logical place to to go about this would be to extend the existing filesystem `sync` property to include a `directory` setting.  Of course the downside is that this would result in a compatibility issue with the other openZFS platforms.  Other solutions are definitely possible and should be investigated.

**Comment by ryao** [comment]
I do not think this is going to pose much of a compatibility issue. We would just need to modify the ZPL bits to pass to the functions from illumos that they are to force ZIL to have sync=always behavior. We probably should ask whether `sync=disabled` or `dirsync` takes precedence. Resolving conflicts by making `sync=disabled` take precedence makes more sense to me than making `dirsync` take precedence. We probably should also restrict the scope of `dirsync` to `sync=standard`, even though `sync=always` and `dirsync` would mean the same thing on directories. That way any new sync settings would be unaffected until someone takes the time to define what `dirsync` means there.

**Comment by behlendorf** [comment]
> The sync property controls how synchronous behavior works and not where synchronous behavior occurs.

I don't agree with this characterization.  For example, when `sync=always` is set it causes normal POSIX operations to be synchronous.  Specifically, `zil_commit()` will be called.  A `sync=directory` version of this would basically be identical to the existing `always` behavior except it would exclude things like `write`.

That said, I'm not a huge fan of extending the `sync` property for the previously mentioned complications and long standing expectations of what it means.  A dedicated `dirsync` property is probably a better approach.  It could be handled similar to something like `relatime`.

> I cannot imagine a sane use case where a system administrator would need to force synchronous behavior on directories and not files.

There's not too much discussion is the original lkml thread about their use case at the time for adding it.  But there are some legitimate use cases for distributed systems where it's critical that directory updates are made synchronously, say to ensure a given ordering.  File updates may not need to be as strict for certain I/O models and making this reasonable compromise to maintain performance.

https://lkml.org/lkml/2002/5/26/106

Besides, since this is a standard Linux mount option it really is something we should support.  Even if it's not critical functionality for most people.
